module_warn.lua

```diff
@@ -265,25 +372,101 @@ function Module:OnLoaded()
 
         Help = "Clears all the warns of a specified user.",
         Silent = true,
         Func = function (commandMessage, targetUser)
             local guild = commandMessage.guild
             local history = self:GetPersistentData(guild)
             local targetMember = guild:getMember(targetUser)
+            local moderator = commandMessage.author
 
             local memberHistory = FindMember(history, targetMember.id)
             if not memberHistory then
                 commandMessage:reply(string.format("The member **%s** (%d) already have zero warns.", targetMember.tag, targetMember.id))
             else
+                for _i, warn in ipairs(memberHistory.Warns) do
+                    self:LogWarnModification(
+                        guild, 
+                        moderator, 
+                        targetMember, 
+                        string.format("**%s** cleared the following warn of **%s** (%d).\nIt was: **%s**\n\t*From: %s*", 
+                            moderator.tag, 
+                            targetMember.tag,
+                            targetMember.id,
+                            warn.Reason,
+                            guild:getMember(warn.From).tag
+                        )
+                    )
+                end
+                
+                self:LogWarnModification(
+                    guild, 
+                    moderator, 
+                    targetMember, 
+                    string.format("**%s** cleared %d warns of **%s** (%d).", 
+                        moderator.tag, 
+                        #memberHistory.Warns,
+                        targetMember.tag,
+                        targetMember.id
+                    )
+                )
+
                 memberHistory.Warns = {}
+
                 commandMessage:reply(string.format("Cleared **%s** (%d) warns, saving.", targetMember.tag, targetMember.id))
                 bot:Save()
             end
         end
     })
 
+    --
+    --  popwarn command
+    --
+    self:RegisterCommand({
+        Name = "popwarn",
+        Args = {
+            {Name = "targetUser", Type = bot.ConfigType.User}
+        },
+        PrivilegeCheck = function (member)
+            return self:HasUnwarnRole(member)
+        end,
+
+        Help = "Removes the last warn of the given user.",
+        Silent = true,
+        Func = function (commandMessage, targetUser)
+            local guild = commandMessage.guild
+            local history = self:GetPersistentData(guild)
+            local targetMember = guild:getMember(targetUser)
+            local moderator = commandMessage.author
+
+            local memberHistory = FindMember(history, targetMember.id)
+            if not memberHistory then
+                commandMessage:reply(string.format("The member **%s** (%d) already have zero warns.", targetMember.tag, targetMember.id))
+            else
+                local lastWarn = table.remove(memberHistory.Warns, #memberHistory.Warns)
```

Comment:
No need for the second table.remove parameter

Reply:
done